pkg/controllers/csi/metadata/sqlite_gorm_client.go

```diff
@@ -0,0 +1,197 @@
+package metadata
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-gormigrate/gormigrate/v2"
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+	"gorm.io/gorm/logger"
+)
+
+type DBAccess interface {
+	SchemaMigration(ctx context.Context) error
+
+	CreateTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error
+	UpdateTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error
+	DeleteTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error
+	ReadTenantConfigByTenantUUID(ctx context.Context, uid string) (*TenantConfig, error)
+
+	CreateCodeModule(ctx context.Context, codeModule *CodeModule) error
+	ReadCodeModuleByVersion(ctx context.Context, version string) (*CodeModule, error)
+	DeleteCodeModule(ctx context.Context, codeModule *CodeModule) error
+
+	CreateOSMount(ctx context.Context, osMount *OSMount) error
+	UpdateOSMount(ctx context.Context, osMount *OSMount) error
+	DeleteOSMount(ctx context.Context, osMount *OSMount) error
+	ReadOSMountByTenantUUID(ctx context.Context, tenantUUID string) (*OSMount, error)
+
+	CreateAppMount(ctx context.Context, appMount *AppMount) error
+	UpdateAppMount(ctx context.Context, appMount *AppMount) error
+	DeleteAppMount(ctx context.Context, appMount *AppMount) error
+	ReadAppMountByVolumeMetaID(ctx context.Context, volumeMetaID string) (*AppMount, error)
+	ReadAppMounts(ctx context.Context) ([]AppMount, error)
+}
+
+type DBConn struct {
+	db *gorm.DB
+}
+
+var _ DBAccess = (*DBConn)(nil)
+
+// NewDBAccess creates a new gorm db connection to the database.
+func NewDBAccess(path string) (DBConn, error) {
+	// we should explicitly enable foreign_keys for sqlite
+	if strings.Contains(path, "?") {
+		path += "&_foreign_keys=on"
+	} else {
+		path += "?_foreign_keys=on"
+	}
+
+	db, err := gorm.Open(sqlite.Open(path), &gorm.Config{Logger: logger.Default})
+
+	if err != nil {
+		return DBConn{}, err
+	}
+
+	return DBConn{db: db}, nil
+}
+
+// SchemaMigration runs gormigrate migrations to create tables
+func (conn *DBConn) SchemaMigration(_ context.Context) error {
+	m := gormigrate.New(conn.db, gormigrate.DefaultOptions, []*gormigrate.Migration{})
+	m.InitSchema(func(tx *gorm.DB) error {
+		err := tx.AutoMigrate(
+			&TenantConfig{},
+			&CodeModule{},
+			&OSMount{},
+			&AppMount{},
+			&VolumeMeta{},
+		)
+		if err != nil {
+			return err
+		}
+		// all other constraints, indexes, etc...
+		return nil
+	})
+
+	_ = m.Migrate()
```

Comment:
Is it a clever idea to add an auto migration before the actual migration, why not having all the migrationsteps in one array?

The problem I see here is that somewhen in the future as soon as we try a migration, the automigrate step could interfere with the predefined migration steps

@andriisoldatenko

Reply:
@gkrenn I didn't get what you mean? Could you please provide more details?

**Context why it's written this way:**
go-migrate strangely behaves if we combine Init migrations and just regular ones. So we decided with @0sewa0 to explicitly migrate the Initial schema, and then you put your migrations after line 81.

So, the new migration will look like it; you should copy the following:
```
gormigrate.New(conn.db, gormigrate.DefaultOptions, []*gormigrate.Migration{
		{
			ID:      "X",
			Migrate: <new migrations here>,
			Rollback: func(tx *gorm.DB) error {
				return nil
			},
		},
	}).Migrate()
```
Below line 81, and you're done. You never touch initial migration.

Reply:
Ok thx, that's what I meant. - Interesting, if there are sideeffects if we combine init migrations and regular ones 🤔

Reply:
If you check my original PR very first few commits I did this way. It creates two records in migration table but never apply second migration lol. Then we spent some time to debug library to understand how it's working under the hood.